### PR TITLE
IOS-7624 Add bottom sheet component in Mistica and Catalog

### PR DESCRIPTION
**Pull request by alejandroruizponce**
## 🎟️ **Jira ticket**
https://jira.tid.es/browse/IOS-7624

## 🥅 **What's the goal?**
Add a generic component to display a bottom sheet selector. At the moment the needs of this bottom sheet are the following:
- It is only a list with rows with text and radioButton.
- It is auto-submittable by clicking on one of the choices or dismisse it.
- It is a single selection.

More information in this Figma link:
https://www.figma.com/file/nxToAYk7FGC86TS0mY61TU/%5BWIP%5D-Sheet?node-id=804%3A13588

In next iterations the idea is to extend it to be able to receive more types of content like buttons, or more than one list. For now the component must be basic and contemplate a single scenario.

## 🚧 **How do we do it?**
The following has been implemented for this purpose:
- BottomSheetView for visual layouts and BottomSheetViewController to display it.
- SheetTransitionDelegate and SheetPresentation for the way to display it on the screen, using fluid transition.
- Added to the catalog for testing.

## 🧪 **How can I verify this?**
Use Mistica Catalog and select the Bottom Sheet component for testing.

### CODE REVIEW

**Comment by yceballost** [comment]
Can you release a version of the catalog to try the component and review it? thanks!

File: Sources/Mistica/Components/BottomSheet/BottomSheet/SheetPresentationController.swift [hunk 1]
```diff
@@ -0,0 +1,286 @@
...
+            handleEndedInteraction()
+
+        // swiftlint:disable:next unnecessary_case_break
+        case .possible:
+            break
+
+        @unknown default:
+            break
+        }
+    }
```
Comment (alexanegon) [inline Sources/Mistica/Components/BottomSheet/BottomSheet/SheetPresentationController.swift:257]: Maybe we can add private functions to a private extension. WDYT?

File: Sources/Mistica/Components/BottomSheet/BottomSheet/SheetTransition.swift [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    // This will get called when the transition is interactive.
+    func interruptibleAnimator(using transitionContext: UIViewControllerContextTransitioning) -> UIViewImplicitlyAnimating {
+        if isPresenting {
+            return presentationAnimator ?? presentationInterruptibleAnimator(using: transitionContext)
+        } else {
+            return dismissAnimator ?? dismissInterruptibleAnimator(using: transitionContext)
+        }
+    }
+
+    private func presentationInterruptibleAnimator(using transitionContext: UIViewControllerContextTransitioning) -> UIViewImplicitlyAnimating {
```
Comment (alexanegon) [inline Sources/Mistica/Components/BottomSheet/BottomSheet/SheetTransition.swift:54]: I would separate private functions into a private extension

File: Sources/Mistica/Components/BottomSheet/Model/BottomSheetSelectionResponse.swift [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+//  BottomSheetSelectionResponse.swift
+//
+//  Made with ❤️ by Novum
+//
+//  Copyright © Telefonica. All rights reserved.
+//
+
+import Foundation
+
+public enum BottomSheetActionResponse: String {
```
Comment (alexanegon) [inline Sources/Mistica/Components/BottomSheet/Model/BottomSheetSelectionResponse.swift:11]: Which are the differences between dismissing and submit?

**Review by alexanegon — COMMENTED** [review]
Issues founded:

1. In the Figma file the asset is aligned with the top of the title.
2. In the catalog app, could we make the description of the elements optional to see the result? It seems that the 24px margin between the title and the row is not like in the Figma section "Contenido configurable"
 

3. Check constraints

`2022-09-13 11:21:11.952379+0200 MisticaCatalog[10592:153084] [LayoutConstraints] Unable to simultaneously satisfy constraints.
	Probably at least one of the constraints in the following list is one you don't want. 
	Try this: 
		(1) look at each constraint and try to figure out which you don't expect; 
		(2) find the code that added the unwanted constraint or constraints and fix it. 
(
    "<NSLayoutConstraint:0x6000010e84b0 V:[UILabel:0x160fb9120]-(16)-[UIStackView:0x160f17500]   (active)>",
    "<NSLayoutConstraint:0x600001318190 'UISV-canvas-connection' V:[UILabel:0x160fb9120]-(0)-|   (active, names: '|':UIStackView:0x160fc7d10 )>",
    "<NSLayoutConstraint:0x600001305630 'UISV-spacing' V:[UIStackView:0x160fc7d10]-(0)-[UIStackView:0x160f17500]   (active)>"
)

Will attempt to recover by breaking constraint 
<NSLayoutConstraint:0x600001318190 'UISV-canvas-connection' V:[UILabel:0x160fb9120]-(0)-|   (active, names: '|':UIStackView:0x160fc7d10 )>

Make a symbolic breakpoint at UIViewAlertForUnsatisfiableConstraints to catch this in the debugger.
The methods in the UIConstraintBasedLayoutDebugging category on UIView listed in <UIKitCore/UIView.h> may also be helpful.`

File: Sources/Mistica/Components/BottomSheet/Model/BottomSheetSelectionResponse.swift (hunk 3 above)
Comment (alejandroruizponce) [reply Sources/Mistica/Components/BottomSheet/Model/BottomSheetSelectionResponse.swift:11]: DISMISS is when you tap outside for example but you didn't select any option. You must send anyway the initial selected option as response. And submit, for now, is when you select an option manually and the sheet is dismissed.

File: Sources/Mistica/Components/BottomSheet/BottomSheet/SheetPresentationController.swift (hunk 1 above)
Comment (alejandroruizponce) [reply Sources/Mistica/Components/BottomSheet/BottomSheet/SheetPresentationController.swift]: Yes, fixed!

File: Sources/Mistica/Components/BottomSheet/BottomSheet/SheetTransition.swift (hunk 2 above)
Comment (alejandroruizponce) [reply Sources/Mistica/Components/BottomSheet/BottomSheet/SheetTransition.swift]: Yes, fixed!

File: Sources/Mistica/Components/BottomSheet/View/BottomSheetView.swift [hunk 4]
```diff
@@ -0,0 +1,250 @@
+//
```
Comment (yceballost) [inline Sources/Mistica/Components/BottomSheet/View/BottomSheetView.swift:1]: Design review
https://telefonicacorp-my.sharepoint.com/:w:/g/personal/ignacio_ceballostorres_telefonica_com/ETSEHJLlQChPv4cXz9-4CnYBoqLx3xZ6NxjpIKinK3UQ6Q?e=qYd3F9

File: Sources/Mistica/Components/BottomSheet/BottomSheet/SheetTransition.swift [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+    var isPresenting = true
+
+    /// The interactive animator used for the dismiss transition.
+    private var dismissAnimator: UIViewPropertyAnimator?
+
+    /// The animator used for the presentation animation.
+    private var presentationAnimator: UIViewPropertyAnimator?
+
+    /// The duration of the transition animation.
+    private let animationDuration: TimeInterval = 0.75
```
Comment (yceballost) [inline Sources/Mistica/Components/BottomSheet/BottomSheet/SheetTransition.swift:25]:
```suggestion
    private let animationDuration: TimeInterval = 0.5
```

**Review by yceballost — CHANGES_REQUESTED** [review]

File: Sources/Mistica/Components/BottomSheet/Model/BottomSheetSelectionResponse.swift (hunk 3 above)
Comment (alexanegon) [reply Sources/Mistica/Components/BottomSheet/Model/BottomSheetSelectionResponse.swift:11]: Nice, maybe we can set this explanation as a comment of the enum. WDYT?

**Comment by jmpg93** [comment]
Hi! I'm missing a few points of this list: https://github.com/Telefonica/mistica-ios/blob/main/CONTRIBUTING.md#how-to-work-with-mistica-ios like tests or documentation.

Also, did you consider accessibility?

**Review by yceballost — DISMISSED** [review]
brilliant ✨

File: Sources/Mistica/Components/Sheet/SheetTransition/SheetPresentationController.swift [hunk 6]
```diff
@@ -0,0 +1,288 @@
...
+    // MARK: - Methods
+
+    override func presentationTransitionWillBegin() {
+        super.presentationTransitionWillBegin()
+
+        containerView?.addSubview(overlayView)
+
+        presentedView?.addSubview(handleView)
+
+        overlayView.fade(toAlpha: 0.5, duration: 0.25)
```
Comment (yceballost) [inline Sources/Mistica/Components/Sheet/SheetTransition/SheetPresentationController.swift:101]:
```suggestion
        overlayView.fade(toAlpha: 1, duration: 0.25)
```

`backgroundOverlay` has its own opacity I think you are applying opacity over the `backgroundOverlay` opacity, I think this value has to be `1` (or maybe remove?)

https://github.com/Telefonica/mistica-ios/blob/4d64130039f83128b8305012ce35575294c720a6/Sources/MisticaCommon/Colors/MovistarColorPalette.swift#L19

**Review by yceballost — CHANGES_REQUESTED** [review]

File: Sources/Mistica/Components/Sheet/View/SheetView.swift [hunk 7]
```diff
@@ -0,0 +1,256 @@
...
+            wrapperView.bottomAnchor.constraint(equalTo: scrollView.bottomAnchor),
+
+            contentStackView.topAnchor.constraint(equalTo: wrapperView.topAnchor),
+            contentStackView.leadingAnchor.constraint(equalTo: wrapperView.leadingAnchor),
+            contentStackView.trailingAnchor.constraint(equalTo: wrapperView.trailingAnchor),
+            contentStackView.bottomAnchor.constraint(equalTo: wrapperView.bottomAnchor, constant: -8.0),
+            contentStackView.widthAnchor.constraint(equalTo: scrollView.widthAnchor),
+
+            itemsStackView.widthAnchor.constraint(equalTo: contentStackView.widthAnchor)
+        ])
```
Comment (alexanegon) [inline Sources/Mistica/Components/Sheet/View/SheetView.swift:238]: Please, check the constraints since the console is warning us:

2022-09-16 08:03:46.894152+0200 MisticaCatalog[73184:2307474] [LayoutConstraints] Unable to simultaneously satisfy constraints.
	Probably at least one of the constraints in the following list is one you don't want. 
	Try this: 
		(1) look at each constraint and try to figure out which you don't expect; 
		(2) find the code that added the unwanted constraint or constraints and fix it. 
(
    "<NSLayoutConstraint:0x6000003863a0 UIStackView:0x14ee64130.width == UIStackView:0x14ee62730.width   (active)>",
    "<NSLayoutConstraint:0x6000003b4960 'UISV-alignment' UIStackView:0x14ee628c0.leading == UIStackView:0x14ee63fa0.leading   (active)>",
    "<NSLayoutConstraint:0x6000003bc780 'UISV-canvas-connection' UIStackView:0x14ee63fa0.leading == UIStackView:0x14ee64130.leading   (active)>",
    "<NSLayoutConstraint:0x6000003bc6e0 'UISV-canvas-connection' H:[Mistica.RadioButton:0x14ee64d30]-(0)-|   (active, names: '|':UIStackView:0x14ee63fa0 )>",
    "<NSLayoutConstraint:0x6000003b4370 'UISV-canvas-connection' UIStackView:0x14ee62730.leading == UIStackView:0x14ee628c0.leading   (active)>",
    "<NSLayoutConstraint:0x6000003b4820 'UISV-canvas-connection' H:[_UILayoutSpacer:0x600001fcb0c0'UISV-alignment-spanner']-(0)-|   (active, names: '|':UIStackView:0x14ee62730 )>",
    "<NSLayoutConstraint:0x6000003bd220 'UISV-spacing' H:[UIStackView:0x14ee64130]-(16)-[Mistica.RadioButton:0x14ee64d30]   (active)>",
    "<NSLayoutConstraint:0x6000003b45f0 'UISV-spanning-boundary' _UILayoutSpacer:0x600001fcb0c0'UISV-alignment-spanner'.trailing >= UIStackView:0x14ee63fa0.trailing   (active)>"
)

Will attempt to recover by breaking constraint 
<NSLayoutConstraint:0x6000003bd220 'UISV-spacing' H:[UIStackView:0x14ee64130]-(16)-[Mistica.RadioButton:0x14ee64d30]   (active)>

Make a symbolic breakpoint at UIViewAlertForUnsatisfiableConstraints to catch this in the debugger.
The methods in the UIConstraintBasedLayoutDebugging category on UIView listed in <UIKitCore/UIView.h> may also be helpful.
2022-09-16 08:03:46.895080+0200 MisticaCatalog[73184:2307474] [LayoutConstraints] Unable to simultaneously satisfy constraints.
	Probably at least one of the constraints in the following list is one you don't want. 
	Try this: 
		(1) look at each constraint and try to figure out which you don't expect; 
		(2) find the code that added the unwanted constraint or constraints and fix it. 
(
    "<NSLayoutConstraint:0x6000003f2940 UIStackView:0x14ee62a50.width == UIStackView:0x14ee62730.width   (active)>",
    "<NSLayoutConstraint:0x6000003a9900 'UISV-canvas-connection' UIStackView:0x14ee628c0.leading == UIStackView:0x14ee62a50.leading   (active)>",
    "<NSLayoutConstraint:0x6000003a9ea0 'UISV-canvas-connection' H:[Mistica.RadioButton:0x14ee63650]-(0)-|   (active, names: '|':UIStackView:0x14ee628c0 )>",
    "<NSLayoutConstraint:0x6000003b4370 'UISV-canvas-connection' UIStackView:0x14ee62730.leading == UIStackView:0x14ee628c0.leading   (active)>",
    "<NSLayoutConstraint:0x6000003b4820 'UISV-canvas-connection' H:[_UILayoutSpacer:0x600001fcb0c0'UISV-alignment-spanner']-(0)-|   (active, names: '|':UIStackView:0x14ee62730 )>",
    "<NSLayoutConstraint:0x6000003aa300 'UISV-spacing' H:[UIStackView:0x14ee62a50]-(16)-[Mistica.RadioButton:0x14ee63650]   (active)>",
    "<NSLayoutConstraint:0x6000003b44b0 'UISV-spanning-boundary' _UILayoutSpacer:0x600001fcb0c0'UISV-alignment-spanner'.trailing >= UIStackView:0x14ee628c0.trailing   (active)>"
)

Will attempt to recover by breaking constraint 
<NSLayoutConstraint:0x6000003aa300 'UISV-spacing' H:[UIStackView:0x14ee62a50]-(16)-[Mistica.RadioButton:0x14ee63650]   (active)>

Make a symbolic breakpoint at UIViewAlertForUnsatisfiableConstraints to catch this in the debugger.
The methods in the UIConstraintBasedLayoutDebugging category on UIView listed in <UIKitCore/UIView.h> may also be helpful.
2022-09-16 08:03:46.914017+0200 MisticaCatalog[73184:2307474] [LayoutConstraints] Unable to simultaneously satisfy constraints.
	Probably at least one of the constraints in the following list is one you don't want. 
	Try this: 
		(1) look at each constraint and try to figure out which you don't expect; 
		(2) find the code that added the unwanted constraint or constraints and fix it. 
(
    "<NSLayoutConstraint:0x60000039c230 UIStackView:0x14ee65610.width == UIStackView:0x14ee62730.width   (active)>",
    "<NSLayoutConstraint:0x6000003b4aa0 'UISV-alignment' UIStackView:0x14ee628c0.leading == UIStackView:0x14ee65480.leading   (active)>",
    "<NSLayoutConstraint:0x6000003b9cc0 'UISV-canvas-connection' UIStackView:0x14ee65480.leading == UIStackView:0x14ee65610.leading   (active)>",
    "<NSLayoutConstraint:0x6000003b9c70 'UISV-canvas-connection' H:[Mistica.RadioButton:0x14ee66210]-(0)-|   (active, names: '|':UIStackView:0x14ee65480 )>",
    "<NSLayoutConstraint:0x6000003b4370 'UISV-canvas-connection' UIStackView:0x14ee62730.leading == UIStackView:0x14ee628c0.leading   (active)>",
    "<NSLayoutConstraint:0x6000003b4820 'UISV-canvas-connection' H:[_UILayoutSpacer:0x600001fcb0c0'UISV-alignment-spanner']-(0)-|   (active, names: '|':UIStackView:0x14ee62730 )>",
    "<NSLayoutConstraint:0x6000003b9bd0 'UISV-spacing' H:[UIStackView:0x14ee65610]-(16)-[Mistica.RadioButton:0x14ee66210]   (active)>",
    "<NSLayoutConstraint:0x6000003b4780 'UISV-spanning-boundary' _UILayoutSpacer:0x600001fcb0c0'UISV-alignment-spanner'.trailing >= UIStackView:0x14ee65480.trailing   (active)>"
)

Will attempt to recover by breaking constraint 
<NSLayoutConstraint:0x6000003b9bd0 'UISV-spacing' H:[UIStackView:0x14ee65610]-(16)-[Mistica.RadioButton:0x14ee66210]   (active)>

Make a symbolic breakpoint at UIViewAlertForUnsatisfiableConstraints to catch this in the debugger.
The methods in the UIConstraintBasedLayoutDebugging category on UIView listed in <UIKitCore/UIView.h> may also be helpful.

File: MisticaCatalog/Source/Catalog/Mistica/Components/UICatalogSheetViewController.swift [hunk 8]
```diff
@@ -0,0 +1,211 @@
...
+private enum Section: Int, CaseIterable {
+    case title
+    case message
+    case description
+    case numberOfElements
+    case hasAsset
+    case showSheet
+}
+
+class UICatalogSheetViewController: UIViewController {
```
Comment (alexanegon) [inline MisticaCatalog/Source/Catalog/Mistica/Components/UICatalogSheetViewController.swift:21]: Same as you added the capability to show or hide the asset in the items, I would add the capability to hide the description.

File: Sources/Mistica/Components/Sheet/SheetTransition/SheetPresentationController.swift (hunk 6 above)
Comment (alejandroruizponce) [reply Sources/Mistica/Components/Sheet/SheetTransition/SheetPresentationController.swift]: Yes, you're right, fixed!

File: MisticaCatalog/Source/Assets.xcassets/Rows/Sheets.imageset/Contents.json [hunk 9]
```diff
@@ -1,7 +1,7 @@
 {
   "images" : [
     {
-      "filename" : "BottomSheets.pdf",
+      "filename" : "Bottom Sheets.svg",
```
[the remark this answers is not in the document]
Comment (alejandroruizponce) [reply MisticaCatalog/Source/Assets.xcassets/Rows/Sheets.imageset/Contents.json]: Fixed!

File: MisticaCatalog/Source/Catalog/Mistica/Components/UICatalogSheetViewController.swift [hunk 10]
```diff
@@ -0,0 +1,211 @@
...
+
+        var rows: [SheetListRow] = []
+        for index in 1 ... sheetNumElements {
+            rows.append(.init(
+                id: index.description,
+                title: "Element \(index)",
+                description: "Description",
+                icon: assetCell.segmentedControl.selectedSegmentIndex == 0 ? .imageIcon : nil
+            ))
+        }
```
[the remark this answers is not in the document]
Comment (alejandroruizponce) [reply MisticaCatalog/Source/Catalog/Mistica/Components/UICatalogSheetViewController.swift]: Fixed!

File: MisticaCatalog/Source/Catalog/Mistica/Components/UICatalogSheetViewController.swift [hunk 11]
```diff
@@ -0,0 +1,211 @@
...
+        )
+
+        let configuration = SheetConfiguration(
+            header: .init(
+                title: sheetTitle,
+                subtitle: sheetSubtitle,
+                description: sheetDescription
+            ),
+            content: [content]
+        )
```
[the remark this answers is not in the document]
Comment (alejandroruizponce) [reply MisticaCatalog/Source/Catalog/Mistica/Components/UICatalogSheetViewController.swift]: Ok, Fixed!

File: MisticaCatalog/Source/Catalog/Mistica/Components/UICatalogSheetViewController.swift [hunk 12]
```diff
@@ -0,0 +1,211 @@
...
+        var rows: [SheetListRow] = []
+        for index in 1 ... sheetNumElements {
+            rows.append(.init(
+                id: index.description,
+                title: "Element \(index)",
+                description: "Description",
+                icon: assetCell.segmentedControl.selectedSegmentIndex == 0 ? .imageIcon : nil
+            ))
+        }
+        let content: SheetList = .init(
```
[the remark this answers is not in the document]
Comment (alejandroruizponce) [reply MisticaCatalog/Source/Catalog/Mistica/Components/UICatalogSheetViewController.swift]: Fixed!

File: MisticaCatalog/Source/Catalog/Mistica/Components/UICatalogSheetViewController.swift [hunk 13]
```diff
@@ -0,0 +1,211 @@
...
+        let configuration = SheetConfiguration(
+            header: .init(
+                title: sheetTitle,
+                subtitle: sheetSubtitle,
+                description: sheetDescription
+            ),
+            content: [content]
+        )
+
+        let viewController = SheetViewController(configuration: configuration) { _ in }
```
[the remark this answers is not in the document]
Comment (alejandroruizponce) [reply MisticaCatalog/Source/Catalog/Mistica/Components/UICatalogSheetViewController.swift]: Ok, added!

File: MisticaCatalog/Source/Catalog/Mistica/Components/UICatalogSheetViewController.swift [hunk 14]
```diff
@@ -0,0 +1,211 @@
...
+
+        present(viewController, animated: true)
+    }
+}
+
+private extension UICatalogSheetViewController {
+    var sheetTitle: String? {
+        guard let text = titleCell.textField.text, !text.isEmpty else { return nil }
+        return text
+    }
```
[the remark this answers is not in the document]
Comment (alejandroruizponce) [reply MisticaCatalog/Source/Catalog/Mistica/Components/UICatalogSheetViewController.swift]: Nope, removed.

File: MisticaCatalog/Source/Common/Views/UIStepperTableViewCell.swift [hunk 15]
```diff
@@ -0,0 +1,71 @@
...
+import Foundation
+import UIKit
+
+public class UIStepperTableViewCell: UITableViewCell {
+    private let numberOfStepsStepper = UIStepper()
+    private let titleLabel = UILabel()
+    private let numberOfStepsLabel = UILabel()
+    private lazy var textStackView = UIStackView(arrangedSubviews: [titleLabel, numberOfStepsLabel])
+    private lazy var numberOfStepsStackView = UIStackView(arrangedSubviews: [textStackView, numberOfStepsStepper])
+    public var didValueChange: ((UIStepper) -> Void)?
```
[the remark this answers is not in the document]
Comment (alejandroruizponce) [reply MisticaCatalog/Source/Common/Views/UIStepperTableViewCell.swift]: Removed!

File: MisticaCatalog/Source/Common/Views/UIStepperTableViewCell.swift [hunk 16]
```diff
@@ -0,0 +1,71 @@
...
+
+    public var title: String? {
+        get { titleLabel.text }
+        set { titleLabel.text = newValue }
+    }
+
+    public var numberOfSteps: String? {
+        get { numberOfStepsLabel.text }
+        set { numberOfStepsLabel.text = newValue }
+    }
```
[the remark this answers is not in the document]
Comment (alejandroruizponce) [reply MisticaCatalog/Source/Common/Views/UIStepperTableViewCell.swift]: Fixed!

File: Sources/Mistica/Components/Sheet/Model/SheetConfiguration.swift [hunk 17]
```diff
@@ -0,0 +1,78 @@
...
+//
+//  Copyright © Telefonica. All rights reserved.
+//
+
+import Foundation
+import UIKit
+
+public struct SheetConfiguration {
+    public var header: SheetHeader
+    public var content: [SheetList]
```
[the remark this answers is not in the document]
Comment (alejandroruizponce) [reply Sources/Mistica/Components/Sheet/Model/SheetConfiguration.swift]: Fixed!

File: Sources/Mistica/Components/Sheet/Model/SheetConfiguration.swift [hunk 18]
```diff
@@ -0,0 +1,78 @@
...
+                content: [SheetList]) {
+        self.header = header
+        self.content = content
+    }
+}
+
+public struct SheetHeader {
+    public var title: String?
+    public var subtitle: String?
+    public var description: String?
```
[the remark this answers is not in the document]
Comment (alejandroruizponce) [reply Sources/Mistica/Components/Sheet/Model/SheetConfiguration.swift]: Fixed!

File: Sources/Mistica/Components/Sheet/Model/SheetConfiguration.swift [hunk 19]
```diff
@@ -0,0 +1,78 @@
...
+    }
+}
+
+public struct SheetList {
+    public var id: String
+    public var type: String
+    public var listType: String?
+    public var autoSubmit: Bool?
+    public var selectedId: String?
+    public var items: [SheetListRow]
```
[the remark this answers is not in the document]
Comment (alejandroruizponce) [reply Sources/Mistica/Components/Sheet/Model/SheetConfiguration.swift]: Fixed!

File: Sources/Mistica/Components/Sheet/Model/SheetConfiguration.swift [hunk 20]
```diff
@@ -0,0 +1,78 @@
...
+        self.items = items
+    }
+}
+
+public struct SheetListRow {
+    public var id: String
+    public var title: String?
+    public var description: String?
+    public var icon: UIImage?
+    public var isSelected: Bool
```
[the remark this answers is not in the document]
Comment (alejandroruizponce) [reply Sources/Mistica/Components/Sheet/Model/SheetConfiguration.swift]: Fixed!

File: Sources/Mistica/Components/Sheet/Model/SheetSelectionResponse.swift [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+//  SheetSelectionResponse.swift
+//
+//  Made with ❤️ by Novum
+//
+//  Copyright © Telefonica. All rights reserved.
+//
+
+import Foundation
+
+public enum SheetActionResponse: String {
```
[the remark this answers is not in the document]
Comment (alejandroruizponce) [reply Sources/Mistica/Components/Sheet/Model/SheetSelectionResponse.swift:13]: Commented!

File: Sources/Mistica/Components/Sheet/Model/SheetSelectionResponse.swift [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+    case dismiss = "DISMISS"
+    case submit = "SUBMIT"
+}
+
+public class SheetSelectionResponse {
+    public var action: SheetActionResponse
+    public var selectedIds: [SheetResponseResult]
```
[the remark this answers is not in the document]
Comment (alejandroruizponce) [reply Sources/Mistica/Components/Sheet/Model/SheetSelectionResponse.swift]: Fixed!

File: Sources/Mistica/Components/Sheet/Model/SheetSelectionResponse.swift [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+    public init(action: SheetActionResponse = .dismiss,
+                selectedIds: [SheetResponseResult] = []) {
+        self.action = action
+        self.selectedIds = selectedIds
+    }
+}
+
+public struct SheetResponseResult: Encodable, Equatable {
+    public let id: String
+    public let selected: [String]
```
[the remark this answers is not in the document]
Comment (alejandroruizponce) [reply Sources/Mistica/Components/Sheet/Model/SheetSelectionResponse.swift]: Fixed!

File: Sources/Mistica/Components/Sheet/View/SheetView.swift (hunk 7 above)
Comment (alejandroruizponce) [reply Sources/Mistica/Components/Sheet/View/SheetView.swift:238]: I do not see these logs when displaying the sheet in either Catalog or Novum. There are no warnings in my debug.

File: MisticaCatalog/Source/Catalog/Mistica/Components/UICatalogSheetViewController.swift (hunk 8 above)
Comment (alejandroruizponce) [reply MisticaCatalog/Source/Catalog/Mistica/Components/UICatalogSheetViewController.swift:21]: For the next iteration we will add the config to edit the rows.

**Review by yceballost — APPROVED** [review]

File: MisticaCatalog/Source/Common/Views/UIStepperTableViewCell.swift [hunk 24]
```diff
@@ -0,0 +1,71 @@
...
+//
+//  Made with ❤️ by Novum
+//
+//  Copyright © Telefonica. All rights reserved.
+//
+
+import Foundation
+import UIKit
+
+public class UIStepperTableViewCell: UITableViewCell {
```
Comment (jmpg93) [inline MisticaCatalog/Source/Common/Views/UIStepperTableViewCell.swift:12]: Drop the UI prefix please.

File: Sources/Mistica/Components/Sheet/View/SheetView.swift (hunk 7 above)
Comment (alexanegon) [reply Sources/Mistica/Components/Sheet/View/SheetView.swift:238]: @jmpg93 do you see those logs?

File: Sources/Mistica/Components/Sheet/Model/SheetConfiguration.swift [hunk 25]
```diff
@@ -0,0 +1,78 @@
...
+                content: [SheetList]) {
+        self.header = header
+        self.content = content
+    }
+}
+
+public struct SheetHeader {
+    let title: String?
+    let subtitle: String?
+    let description: String?
```
Comment (jmpg93) [inline Sources/Mistica/Components/Sheet/Model/SheetConfiguration.swift:26]: Are all three texts optional? What if we make the title non-optional and make the whole `Section` optional? So you don't have to build an empty section to create a `SheetConfiguration`?

```
let configuration = SheetConfiguration(header: nil, content: [])
vs
let configuration = SheetConfiguration(header: Section(), content: [])
```

File: Sources/Mistica/Components/Sheet/SheetTransition/SheetPresentationController.swift [hunk 26]
```diff
@@ -0,0 +1,288 @@
...
+    private let dismissThreshold: CGFloat = 0.3
+
+    // MARK: - Computed Properties
+
+    override var frameOfPresentedViewInContainerView: CGRect {
+        guard let containerView = containerView, let presentedView = presentedView else {
+            return super.frameOfPresentedViewInContainerView
+        }
+
+        /// The maximum height allowed for the sheet. We allow the sheet to reach the top safe area inset.
```
Comment (jmpg93) [inline Sources/Mistica/Components/Sheet/SheetTransition/SheetPresentationController.swift:28]: Comments that are not documentation should use `//` instead of `///`

File: Sources/Mistica/Components/Sheet/View/SheetView.swift [hunk 27]
```diff
@@ -0,0 +1,256 @@
...
+        let label = UILabel()
+        label.text = config.header.description
+        label.numberOfLines = 0
+        label.textColor = .textSecondary
+        label.font = .textPreset2(weight: .regular)
+        return label
+    }()
+
+    private let config: SheetConfiguration
+    private var contentInfo: [String: [String: RadioButton]] = [:]
```
Comment (jmpg93) [inline Sources/Mistica/Components/Sheet/View/SheetView.swift:41]: What is this content info? Can't we use a struct os something like that?

**Review by alexanegon — DISMISSED** [review]

**Review by yceballost — APPROVED** [review]

File: Sources/Mistica/Components/Sheet/Model/SheetSelectionResponse.swift [hunk 28]
```diff
@@ -0,0 +1,38 @@
...
+//  Copyright © Telefonica. All rights reserved.
+//
+
+import Foundation
+
+// - DISMISS: when you tap outside for example but you didn't select any option. You must send anyway the initial selected option as response.
+// - SUBMIT when you select an option manually and the sheet is dismissed.
+public enum SheetActionResponse: String {
+    case dismiss = "DISMISS"
+    case submit = "SUBMIT"
```
Comment (jmpg93) [inline Sources/Mistica/Components/Sheet/Model/SheetSelectionResponse.swift:15]: Do you really need this to have a string value?

File: Sources/Mistica/Components/Sheet/View/SheetView.swift [hunk 29]
```diff
@@ -0,0 +1,256 @@
...
+    private var contentSelected: [SheetResponseResult] = []
+    public var dismissSheet: (() -> Void)?
```
Comment (jmpg93) [inline Sources/Mistica/Components/Sheet/View/SheetView.swift:43]: I would have made this to return the response.
dismissSheet: (SheetResponse) -> Void

File: Sources/Mistica/Components/Sheet/View/SheetView.swift [hunk 30]
```diff
@@ -0,0 +1,256 @@
...
+                itemStackView.axis = .horizontal
+                itemStackView.distribution = .fill
+                itemStackView.alignment = (item.description != nil) ? .top : .center
+                itemStackView.spacing = 16.0
+
+                let itemTapGesture = ItemInformationTapGesture(target: self, action: #selector(selectItem(_:)))
+                itemTapGesture.contentId = content.id
+                itemTapGesture.itemId = item.id
+                itemTapGesture.autoSubmit = content.autoSubmit ?? false
+                itemStackView.addGestureRecognizer(itemTapGesture)
```
Comment (jmpg93) [inline Sources/Mistica/Components/Sheet/View/SheetView.swift:131]: Instead of creating all the elements on appear, we could have use a list with reusable Cells. A mistica Cell can be easily configured as the rows of the sheet.

File: Sources/Mistica/Components/Sheet/View/SheetView.swift [hunk 31]
```diff
@@ -0,0 +1,256 @@
...
+                }
+
+                itemStackView.addArrangedSubview(elementTextStackView)
+                containerItemStackView.addArrangedSubview(itemStackView)
+
+                let radioButton = RadioButton()
+                radioButton.isActivated = item.id == content.selectedId
+                radioButton.isUserInteractionEnabled = false
+                radioButton.heightAnchor.constraint(equalToConstant: 20.0).isActive = true
+                radioButton.widthAnchor.constraint(equalToConstant: 20.0).isActive = true
```
Comment (jmpg93) [inline Sources/Mistica/Components/Sheet/View/SheetView.swift:175]: I think the radio button doesn't need constraints.

**Comment by jmpg93** [comment]
I'm still missing tests and the SwiftUI implementation.

File: Makefile [hunk 32]
```diff
@@ -24,7 +24,7 @@ XCODEBUILD := set -o pipefail && xcodebuild
 
 # Xcode
 ifneq ($(origin GITHUB_ACTION),undefined)
-export DEVELOPER_DIR=/Applications/Xcode-13.2.app/Contents/Developer
+export DEVELOPER_DIR=/Applications/Xcode-13.4.1.app/Contents/Developer
```
Comment (amegias) [inline Makefile:27]: it's wrong @alejandroruizponce

**Review by amegias — APPROVED** [review]

**Review by amegias — APPROVED** [review]

**Comment by jmpg93** [comment]
Please, if you are going to merge this branch because we need the changes for 13.8, create ticket with the following:
- Implemente SwiftUI version
- Add snapshot test.
- Try to simplify the setup of the sheet: it is hard to follow the view definition with all that constraints.
- Replace doc slashes with comment slashes
